# **Heart Failure Data: Extract, Transform and Load**

## Objectives

* Extract, clean and save the heart failure data.

## Inputs

* Dataset file data/raw/heart.csv

## Outputs

* A cleaned csv file, to be named later.

## Additional Comments

* Wasted at least 30 minutes trying to get into the correct working directory.  My tremor meant that I was clicking, using my trackpad, twice on the white Run icon to the left of the notebook cells and so traversing one too many directories.  Changing to selecting a cell and then running it with Shift + Enter solved the problem.





---

# Change working directory

* We are assuming you will store the notebooks in a subfolder, therefore when running the notebook in the editor, you will need to change the working directory

We need to change the working directory from its current folder to its parent folder
* We access the current directory with os.getcwd()

In [1]:
import os
current_dir = os.getcwd()
current_dir

'c:\\Code Institute - Data Analytics\\heart-failure-analysis\\jupyter_notebooks'

We want to make the parent of the current directory the new current directory
* os.path.dirname() gets the parent directory
* os.chir() defines the new current directory

In [2]:
os.chdir(os.path.dirname(current_dir))
print("You set a new current directory")

You set a new current directory


Confirm the new current directory

In [3]:
current_dir = os.getcwd()
current_dir

'c:\\Code Institute - Data Analytics\\heart-failure-analysis'

# Extract

In [4]:
import pandas as pd, numpy as np, matplotlib.pyplot as plt, seaborn as sns

In [6]:
df = pd.read_csv("data/raw/heart.csv")
df.head()

,Age,Sex,ChestPainType,RestingBP,Cholesterol,FastingBS,RestingECG,MaxHR,ExerciseAngina,Oldpeak,ST_Slope,HeartDisease
0,40,M,ATA,140,289,0,Normal,172,N,0.0,Up,0
1,49,F,NAP,160,180,0,Normal,156,N,1.0,Flat,1
2,37,M,ATA,130,283,0,ST,98,N,0.0,Up,0
3,48,F,ASY,138,214,0,Normal,108,Y,1.5,Flat,1
4,54,M,NAP,150,195,0,Normal,122,N,0.0,Up,0


In [9]:
df.shape

(918, 12)

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 918 entries, 0 to 917
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Age             918 non-null    int64  
 1   Sex             918 non-null    object 
 2   ChestPainType   918 non-null    object 
 3   RestingBP       918 non-null    int64  
 4   Cholesterol     918 non-null    int64  
 5   FastingBS       918 non-null    int64  
 6   RestingECG      918 non-null    object 
 7   MaxHR           918 non-null    int64  
 8   ExerciseAngina  918 non-null    object 
 9   Oldpeak         918 non-null    float64
 10  ST_Slope        918 non-null    object 
 11  HeartDisease    918 non-null    int64  
dtypes: float64(1), int64(6), object(5)
memory usage: 86.2+ KB


In [13]:
df.isnull().sum()

Age               0
Sex               0
ChestPainType     0
RestingBP         0
Cholesterol       0
FastingBS         0
RestingECG        0
MaxHR             0
ExerciseAngina    0
Oldpeak           0
ST_Slope          0
HeartDisease      0
dtype: int64

In [15]:
df.duplicated().sum()

0

The investigation of the heart.csv file shows that it is comprised of 918 rows (labelled from 0 to 917), and 12 columns (labelled from 0 to 11).

The various columns of data can be divided into numeric and non-numeric datatypes (7 and 5 columns respectively).  An alternative categorisation may be made based on other features of the data such as demographics, clinical chemistry laboratory results, or physiological measures of cardiac function, emphasizing the richness of the dataset, despite its small size.

There appear to be no missing values or repeated rows of data.

In an ideal world I would like to see more information on the provenance of the data.  After using the Internet Archive it appears that one or more datasets from the 1990s were the source for the current dataset but the current dataset makes no mention of how it was derived.  I am therefore sceptical with regard to its quality.

#   Column           Dtype        Category            Meaning
--  ------           -----        --------            -------
 0   Age             numeric      Demographics        Age (at diagnosis?)
 1   Sex             text         Demographics        Sex
 2   ChestPainType   text         Cardiac function    Type of chest pain
 3   RestingBP       numeric      Cardiac function    Resting blood pressure
 4   Cholesterol     numeric      Clinical chemistry  Cholesterol level
 5   FastingBS       numeric      Clinical chemistry  Fasting blood sugar
 6   RestingECG      text         Cardiac function    Resting electrocardiogram
 7   MaxHR           numeric      Cardiac function    Maximum heart rate
 8   ExerciseAngina  text         Cardiac function    Angina on exercising
 9   Oldpeak         numeric      Cardiac function    ECG ST segment level
 10  ST_Slope        text         Cardiac function    ST segment slope
 11  HeartDisease    numeric      Cardiac function    Presence of heart disease
Data types: numeric(7), text(5)

In [17]:
df.describe()

,Age,RestingBP,Cholesterol,FastingBS,MaxHR,Oldpeak,HeartDisease
count,918.000000,918.000000,918.000000,918.000000,918.000000,918.000000,918.000000
mean,53.510893,132.396514,198.799564,0.233115,136.809368,0.887364,0.553377
std,9.432617,18.514154,109.384145,0.423046,25.460334,1.066570,0.497414
min,28.000000,0.000000,0.000000,0.000000,60.000000,-2.600000,0.000000
25%,47.000000,120.000000,173.250000,0.000000,120.000000,0.000000,0.000000
50%,54.000000,130.000000,223.000000,0.000000,138.000000,0.600000,1.000000
75%,60.000000,140.000000,267.000000,0.000000,156.000000,1.500000,1.000000
max,77.000000,200.000000,603.000000,1.000000,202.000000,6.200000,1.000000


## Definition of Statistical Terms Used in Describe Table

| Term                     | Meaning                                          |
| :----------------------- | :----------------------------------------------- |
| Mean                     | The average of the values ie sum all the values, <br>and divide the sum by the number of values.      |
| Median                   | The middle value when the values are sorted.     |
| Std (standard deviation) | A measure of how spread out the values are <br>around the mean                               |
| Min/Max                  | The smallest and largest values respectively.    |
| 25%/50%/75%              | The  that splits the sorted data so 25%, <br>50% or 75% of all values fall below it.  They <br>are also known as quartiles.  The 50% quartile<br> is the median.                                   |

In [24]:
(df["Cholesterol"] == 0).sum()

172

In [35]:
df[df["Cholesterol"] == 0]["HeartDisease"].value_counts()

HeartDisease
1    152
0     20
Name: count, dtype: int64

In [36]:
df["HeartDisease"].value_counts()

HeartDisease
1    508
0    410
Name: count, dtype: int64

The 'zero' Cholesterol rows contained 152 patients with HeartDisease (88.4% of the zero-cholesterol group) and 20 patients (11.6%) with no HeartDisease.  This compares with all 918 patients who were split with 508 patients having heart disease (55.3% of the total) and 410 (44.7%) without heart disease.  Thus, patients in the subgroup with a cholesterol value of zero appear to be more likely to have heart disease than patients overall. To remove these patients from the dataset could bias the sample, removing a large number of patients that one would want to study i.e. those patients with heart disease.

In [37]:
(df["RestingBP"] == 0).sum()

1

In [41]:
df_clean = df.copy()
df_clean["Cholesterol_Missing"] = (df_clean["Cholesterol"] == 0).astype(int)
df_clean["Cholesterol_Missing"].sum()

172

In [42]:
df_clean.columns

Index(['Age', 'Sex', 'ChestPainType', 'RestingBP', 'Cholesterol', 'FastingBS',
       'RestingECG', 'MaxHR', 'ExerciseAngina', 'Oldpeak', 'ST_Slope',
       'HeartDisease', 'Cholesterol_Missing'],
      dtype='object')

In [44]:
cholesterol_median = df_clean.loc[df_clean["Cholesterol"] != 0, "Cholesterol"].median()
print(cholesterol_median)

237.0


In [46]:
df_clean["Cholesterol"] = df_clean["Cholesterol"].replace(0, cholesterol_median)

In [ ]:
# Has replacing cholesterol zeroes with median worked as expected?
(df_clean["Cholesterol"] == 0).sum()    #   0
df_clean["Cholesterol_Missing"].sum()   # 172
(df_clean["Cholesterol"] == 237).sum()  # 178
df_clean["Cholesterol"].median()        # 237.0
# Checks after replacing zeroes with median all gave expected results, and revealed that there were six original values of 237.0 prior to the replacement.

237.0

## The Fate of RestingBP

The search for zeros revealed that the vast majority of them occurred within 172 rows, in the Cholesterol field.  However, there was also a single instance of a zero located in the RestingBP field, which had to be erroneous as a resting blood pressure of zero is incompatible with life.  This posed a problem of how it was to be tackled.  The isolated row could either be discarded, with potential loss of useful information, or it could be retained and a value for the resting blood pressure imputed using the median of the non-zero values in the field.

I decided to keep the row and impute a value for resting blood pressure, despite the corresponding value for Cholesterol also having been missing,  with a zero value, originally.  I believe imputation for RestingBP, with the associated retention of other data values in the row, outweighs the argument for removing a partial record with two  erroneous data values.

Following replacement by the median value, 119 instances of the median were found in RestingBP, and the median was 130.  The shape of the dataset at this stage was (918, 13) i.e. all original rows are retained, and an additional 'flag' column, Cholesterol_Missing, has been created.

In [57]:
df_clean[df_clean["RestingBP"] == 0]

,Age,Sex,ChestPainType,RestingBP,Cholesterol,FastingBS,RestingECG,MaxHR,ExerciseAngina,Oldpeak,ST_Slope,HeartDisease,Cholesterol_Missing
449,55,M,NAP,0,237,0,Normal,155,N,1.5,Flat,1,1


In [71]:
resting_BP_median = df_clean.loc[df_clean["RestingBP"] != 0, "RestingBP"].median()
print(resting_BP_median)
df_clean["RestingBP"] = df_clean["RestingBP"].replace(0, resting_BP_median)
# Has replacing resting BP zero with median worked as expected?
#(df_clean["RestingBP"] == 0).sum()   #   0
#(df_clean["RestingBP"] == resting_BP_median).sum()  # 119
#df_clean["RestingBP"].median()        # 130.0
df_clean.shape
# Checks after replacing zeroes with median all gave expected results, and revealed that there were six original values of 237.0 prior to the replacement.

130.0


(918, 13)

---

# Transform

---

NOTE

* You may add as many sections as you want, as long as it supports your project workflow.
* All notebook's cells should be run top-down (you can't create a dynamic wherein a given point you need to go back to a previous cell to execute some task, like go back to a previous cell and refresh a variable content)

---

# Push files to Repo

* In cases where you don't need to push files to Repo, you may replace this section with "Conclusions and Next Steps" and state your conclusions and next steps.

In [45]:
import os
try:
  # create your folder here
  # os.makedirs(name='')
except Exception as e:
  print(e)


IndentationError: expected an indented block after 'try' statement on line 2 (553063055.py, line 5)